File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    List<String> validCombinations = new ArrayList<>();
    Stack<Character> combinationStack = new Stack<>();
    int maxCount = 0;

    public List<String> generateParenthesis(int n) {
        int openCount = 0;
        int closedCount = 0;
        maxCount = n;
        handleCombination(openCount, closedCount);
        return validCombinations;
    }

    private void handleCombination(int openCount, int closedCount) {
        if (openCount < maxCount) {
            combinationStack.push('(');
            handleCombination(openCount + 1, closedCount);
            combinationStack.pop();
        }
        if (closedCount < openCount) {
            combinationStack.push(')');
            handleCombination(openCount, closedCount + 1);
            combinationStack.pop();
        }
        if (openCount == maxCount && closedCount == maxCount) {
            validCombinations.add(stackToString(combinationStack));
        }
    }

    private String stackToString(Stack<Character> stack) {
        StringBuilder sb = new StringBuilder();
        for (char c : stack) {
            sb.append(c);
        }
        return sb.toString();
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py

class Solution:
    def generateParenthesis(self, n: int):
        all_combinations = []
        def generate(n_available, n_open, current_sequence):
            if n_available == 0 and n_open == 0:
                all_combinations.append("".join(sequence))
            if n_available > 0:
                current_sequence.append("(")
                generate(n_available - 1, n_open + 1, current_sequence)
                current_sequence.pop()	
            if n_open > 0:
                current_sequence.append(")") 
                generate(n_available, n_open - 1, current_sequence)
                current_sequence.pop()
        sequence = []
        generate(n, 0, sequence)
        return all_combinations